File: sdk/typescript/plugin-sdk/src/protocol.ts
export const JSON_RPC_VERSION = "2.0";
export const APP_VERSION = "1.0.0";

export const METHODS = {
  init: "init",
  listTasks: "listTasks",
  executeTask: "executeTask",
  shutdown: "shutdown",
} as const;

export const EVENT_TYPES = {
  output: "output",
  progress: "progress",
  error: "error",
  completed: "completed",
} as const;

export type JsonPrimitive = string | number | boolean | null;
export type JsonValue = JsonPrimitive | JsonObject | JsonArray;
export interface JsonObject {
  [key: string]: JsonValue;
}
export type JsonArray = JsonValue[];

export interface JsonRpcRequest<TParams = unknown> {
  jsonrpc: typeof JSON_RPC_VERSION;
  id: number;
  method: string;
  params?: TParams;
}

export interface JsonRpcError {
  code: number;
  message: string;
  data?: JsonValue;
}

export interface JsonRpcResponse<TResult = unknown> {
  jsonrpc: typeof JSON_RPC_VERSION;
  id: number | null;
  result?: TResult;
  error?: JsonRpcError;
}

export interface InitParams {
  config?: JsonObject;
  protocol_version: string;
}

export interface InitResult {
  ok: boolean;
  name?: string;
  version?: string;
}

export interface TaskDescriptor {
  id: string;
  description?: string;
  phase?: string;
  dependencies?: string[];
  permissions?: string[];
  requires_confirmation?: boolean;
}

export interface ExecuteTaskParams {
  id: string;
  args?: string[];
  env?: Record<string, string>;
}

export interface TaskEvent {
  type: (typeof EVENT_TYPES)[keyof typeof EVENT_TYPES];
  data?: string;
  progress?: number;
  exitCode?: number;
}

export const JSON_RPC_ERROR = {
  parseError: -32700,
  invalidRequest: -32600,
  methodNotFound: -32601,
  internalError: -32603,
} as const;

File: sdk/go/plugin-sdk-go/protocol.go
package architectplugin

const (
	JSONRPCVersion = "2.0"
	APPVersion     = "1.0.0"

	MethodInit        = "init"
	MethodListTasks   = "listTasks"
	MethodExecuteTask = "executeTask"
	MethodShutdown    = "shutdown"

	EventOutput    = "output"
	EventProgress  = "progress"
	EventError     = "error"
	EventCompleted = "completed"

	ErrorParseError    = -32700
	ErrorInvalid       = -32600
	ErrorMethodMissing = -32601
	ErrorInternal      = -32603
)

type Request struct {
	JSONRPC string `json:"jsonrpc"`
	ID      int    `json:"id"`
	Method  string `json:"method"`
	Params  any    `json:"params,omitempty"`
}

type Response struct {
	JSONRPC string    `json:"jsonrpc"`
	ID      *int      `json:"id"`
	Result  any       `json:"result,omitempty"`
	Error   *RPCError `json:"error,omitempty"`
}

type RPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    any    `json:"data,omitempty"`
}

type InitParams struct {
	Config          map[string]any `json:"config,omitempty"`
	ProtocolVersion string         `json:"protocol_version"`
}

type InitResult struct {
	OK      bool   `json:"ok"`
	Name    string `json:"name,omitempty"`
	Version string `json:"version,omitempty"`
}

type TaskDescriptor struct {
	ID                   string   `json:"id"`
	Description          string   `json:"description,omitempty"`
	Phase                string   `json:"phase,omitempty"`
	Dependencies         []string `json:"dependencies,omitempty"`
	Permissions          []string `json:"permissions,omitempty"`
	RequiresConfirmation bool     `json:"requires_confirmation,omitempty"`
}

type ExecuteTaskParams struct {
	ID   string            `json:"id"`
	Args []string          `json:"args,omitempty"`
	Env  map[string]string `json:"env,omitempty"`
}

type TaskEvent struct {
	Type     string   `json:"type"`
	Data     string   `json:"data,omitempty"`
	Progress *float64 `json:"progress,omitempty"`
	ExitCode *int     `json:"exitCode,omitempty"`
}
